### PR TITLE
Justin reynolds mm patch 1

### PR DESCRIPTION
**Pull request by JustinReynolds-MM**
Added Performance Alerting Guide.

### CODE REVIEW

**Comment by amyblais** [comment]
@lfbrock For screenshot review.

**Comment by amyblais** [comment]
I missed that - @JustinReynolds-MM can you help re-upload the images via docs -> source -> images. You can ping me for any questions

**Comment by jasonblais** [comment]
@JustinReynolds-MM Can you also help remove the images not in the `images` folder? 
image

I'll try and see if I can get less blurry screenshots for the first few in the doc.

File: source/administration/performance-alerting-guide.rst [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+The `Mattermost dashboards <https://grafana.com/dashboards?search=mattermost>`_ for Grafana come with some partially pre-configured alerts on the following charts:
+
+* CPU Utilization Rate
+* Memory Usage
+* Number of Goroutines
+* Number of API Errors per Second
+* Mean API Request Time
+
+To configure alerts 1) set an appropriate threshold and 2) enable notifications. Enabling notifications are the same for each chart, but setting the correct threshold can have some variances that are better handled on a per-chart basis.
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: Propose

```
To configure alerts, set an appropriate threshold and enable notifications. Enabling notifications is the same for each chart, but setting the correct threshold can have some variances that are better handled on a per-chart basis.
```

File: source/administration/performance-alerting-guide.rst [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+For each chart, click on the chart name and click “Edit”:
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: Let's number the steps, so `1.` here

File: source/administration/performance-alerting-guide.rst [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+.. image:: ../images/perf-1.png
+
+Then click on the “Alert” tab:
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: `2.` here

File: source/administration/performance-alerting-guide.rst [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+.. image:: ../images/perf-2.png
+
+The alert threshold, which will be discussed in the sections below, is the last field under “Conditions” (the one set to 600 in the screenshot above).
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: `3.` here

File: source/administration/performance-alerting-guide.rst [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+See the sections below for how to set the threshold for each individual chart. If you would like to add your own custom alert conditions, configure them here.
+
+.. image:: ../images/perf-3.png
+
+To enable the notifications for any alerts, click on the “Notification” tab on the left and select “Mattermost Alerts Channel” under “Send to”:
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: `4.` here

File: source/administration/performance-alerting-guide.rst [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+.. image:: ../images/perf-4.png
+
+Fill in a message if you would like to add more context to the alert.
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: `Fill in` --> `Enter`

File: source/administration/performance-alerting-guide.rst [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+By default, the alerts are configured to check the average of a chart over the last minute to see if that value is above a threshold. If it’s above the threshold, the alert will be triggered. Since it’s an average over the last minute, small spikes that go past the threshold won’t necessarily cause an alert. This helps prevent false positives that would result from natural spikes in usage. The alert state for each chart is evaluated every minute.
+
+The sections below describe each chart in more detail.
+
+CPU Utilization Rate
+---------------------
+
+CPU Utilization Rate is fairly straightforward. It tracks the CPU usage of the app servers as a percentage. The maximum percentage is based on the number of CPU cores or vCPUs your app server has. For example, if you have four CPU cores and your app server was at 100% utilization rate on all four cores, the graph would show 400% for that app server.
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: `CPU Utilization Rate is fairly straightforward. It tracks...` --> `CPU Utilization Rate tracks...`

File: source/administration/performance-alerting-guide.rst [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+It’s best to set the alert threshold based on your average CPU utilization and how many cores/vCPUs your app servers have. Take a look at the chart over the last seven days. You’ll want to set the threshold somewhere between your maximum CPU usage (cores * 100) and the CPU usage you see. The lower you set the threshold, the more alerts, and therefore the more false positives, you will get. Set the threshold too high and you may not receive an alert for an incident until it’s gotten worse. The same principle applies to all alerts, regardless of the chart.
+
+For example, on our daily build server that much of the team and community works out of, we have the threshold set to 15%:
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: `For example, on our community server, we have the threshold set to 15%:`

File: source/administration/performance-alerting-guide.rst [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+.. image:: ../images/perf-5.png
+
+This is well below our maximum CPU usage as we don’t have a high load. It’s also above our average usage at peak times, so we will get alerts if we start experiencing irregularly high CPU usage.
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: `This value is below our maximum CPU usage and above our average usage at peak times. Therefore, we will get alerts if we begin experiencing unusually high CPU usage.`

File: source/administration/performance-alerting-guide.rst [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+Memory Usage
+-------------
+
+Memory Usage is also straightforward. It involves simply tracking the megabytes of RAM that your app servers are using. You’ll want to set the threshold much like you did for the CPU Utilization Rate: below maximum available memory and above your average usage during peak times.
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: Start with `Memory Usage tracks megabytes of RAM...`

Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]:
```
You’ll want to set the threshold much like you did for the CPU Utilization Rate: below maximum available memory and above your average usage during peak times.
```

to

```
Set the threshold similar to the CPU Utilization Rate: below maximum available memory and above your average usage during peak times.
```

File: source/administration/performance-alerting-guide.rst [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+Here’s how we have the alert set on our daily build server:
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: `daily build server` --> `community server`

File: source/administration/performance-alerting-guide.rst [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+Memory Usage is also straightforward. It involves simply tracking the megabytes of RAM that your app servers are using. You’ll want to set the threshold much like you did for the CPU Utilization Rate: below maximum available memory and above your average usage during peak times.
+
+Here’s how we have the alert set on our daily build server: 
+
+.. image:: ../images/perf-6.png
+
+Number of Goroutines
+---------------------
+
+If you’re not familiar with Go and goroutines, this chart might not make much sense to you upfront. Goroutines are functions or methods that run concurrently with other functions and methods. You can think of them as lightweight threads with low-creation costs. A rising number of goroutines can be a good measure of the performance of your app servers. If they begin to continuously climb, that indicates your app server can’t keep up and is creating goroutines faster than they can complete their tasks and stop.
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: Remove `If you’re not familiar with Go and goroutines, this chart might not make much sense to you upfront.` as it doesn't add value to the guide.

Comment (jasonblais) [reply source/administration/performance-alerting-guide.rst]:
```
You can think of them as lightweight threads with low-creation costs. A rising number of goroutines can be a good measure of the performance of your app servers. If they begin to continuously climb, that indicates your app server can’t keep up and is creating goroutines faster than they can complete their tasks and stop.
```

to

```
Goroutines are like lightweight threads with low-creation costs. A rising number of goroutines can be a good measure of the performance of your app servers. A continuous increase indicates your app server can't keep up and is creating goroutines faster than they can complete their tasks and stop.
```

File: source/administration/performance-alerting-guide.rst [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+Set the threshold somewhere above the average number of goroutines you see during peak load times. Small spikes are usually nothing to worry about. It’s the uncontrolled climbing of goroutines that you want to watch out for.
+
+Here’s how we have it set on our daily build server:
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: `daily build server` --> `community`

File: source/administration/performance-alerting-guide.rst [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+.. image:: ../images/perf-7.png
+
+Number of API Errors per Second
+--------------------------------
+
+Any 4xx or 5xx HTTP response status codes are counted as a REST API error. API errors themselves are not necessarily a problem. There are many legitimate reasons for an API error to occur, such as users’ sessions expiring or clients requesting to see if a resource exists and is being given a “404 Not Found” response. It is normal to have some API errors that scale with your installation base.
+
+That said, errors against the REST API can be indicative of deployment and other issues. For example, if one of your app servers did not deploy correctly for whatever reason, it may begin returning a high number of API errors. Another example would be a rogue bot spamming the API with bad requests. Alerts on API errors per second would help catch these and other issues.
+
+Here’s how it’s set on our daily build server:
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: `daily build server` --> `community`

File: source/administration/performance-alerting-guide.rst [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+.. image:: ../images/perf-8.png
+
+Mean API Request Time
+----------------------
+
+The Mean API Request Time is the average amount of time a REST API request to the Mattermost app server takes to complete. If an app server starts to perform poorly, you’ll likely see a rise in the mean request time as it takes longer to complete requests. This could also happen if your database can’t sustain the load from the app servers. It may also be indicative of an issue between the app servers and your proxy.
+
+You’ll want to set the alert threshold a little above what the mean request time is during your peak load times.
+
+Here’s how it’s set on our daily build server:
```
Comment (jasonblais) [inline source/administration/performance-alerting-guide.rst]: `daily build server` --> `community server`

**Comment by lfbrock** [comment]
Thanks @jasonblais - they look better now, a couple still look mildly blurry but I think it's just the font rendering of some of the fonts on Windows.

**Pull request merged**